File: 16. 3Sum Closest/Solution.cpp
// LeetCode #16: 3Sum Closest

/*
Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

    For example, given array S = {-1 2 1 -4}, and target = 1.

    The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
*/

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int min = INT_MAX; // NOT INT_MIN.
        int res = 0;
        
        sort(nums.begin(), nums.end());
        
        for (int i = 0; i < nums.size(); i++) {
            int j = i+1;
            int k = nums.size() - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                int diff = abs(target - sum);
                
                if (diff < min) {
                    min = diff;
                    res = sum;
                }
                
                if (target > sum)
                    j++;
                else
                    k--;
            }
        }
        return res;
    }
};

File: 16. 3Sum Closest/Solution.py
# LeetCode #16: 3Sum Closest

'''
Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

    For example, given array S = {-1 2 1 -4}, and target = 1.

    The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
'''

class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer}
    def threeSumClosest(self, nums, target):
        minValue = 1 << 32
        res = 0
        
        nums.sort()
        
        for i in xrange(len(nums)):
            j = i+1
            k = len(nums)-1
            while j < k:
                threeSum = nums[i] + nums[j] + nums[k]
                diff = abs(target - threeSum)
                if diff < minValue:
                    minValue = diff
                    res = threeSum
                    
                if target > threeSum:
                    j += 1
                else:
                    k -= 1
        return res